draft-ietf-tls-tls13.md

```diff
@@ -1622,29 +1637,31 @@ Key Exchange: establish shared keying material and select the
    encrypted.
 
 Server Parameters: establish other handshake parameters
-(whether the client is authenticated, support for 0-RTT, etc.)
+(whether the client is authenticated, application layer protocol support, etc.)
 
 Authentication: authenticate the server (and optionally the client)
    and provide key confirmation and handshake integrity.
 
-In the Key Exchange phase, the client sends the ClientHello ({{client-hello}})
-message, which contains
-a random nonce (ClientHello.random), its offered protocol version,
-cipher suite, and extensions, and one or more Diffie-Hellman key
-shares in the "key_share" extension {{key-share}}.
+In the Key Exchange phase, the client sends the ClientHello
+({{client-hello}}) message, which contains a random nonce
+(ClientHello.random), its offered protocol version, cipher suite, and
+extensions, and in general either one or more Diffie-Hellman key shares (in the
+"key_share" extension {{key-share}}), one or more pre-shared key labels (in the
+"pre_shared_key" extension {{pre-shared-key-extension}}), or both.
 
 The server processes the ClientHello and determines the appropriate
 cryptographic parameters for the connection. It then responds with
-its own ServerHello which indicates the negotiated connection parameters. [{{server-hello}}]
-If DH is in use, this will contain a "key_share" extension with the
-server's ephemeral Diffie-Hellman share which MUST be in the same group
-as one of the shares offered by the client. The server's KeyShare and
-the client's KeyShare corresponding to the negotiated key exchange
-are used together to derive the Static Secret and Ephemeral
-Secret (in this mode they are the same).  [{{key-share}}]
-
-
-The server then sends three messages to establish the Server Parameters:
+its own ServerHello which indicates the negotiated connection parameters. [{{server-hello}}].
+The combination of the ClientHello and the ServerHello determines
+the values of ES and SS, as described above. If either a pure
+(EC)DHE or (EC)DHE-PSK cipher suite is in use, then the ServerHello
+will will contain a "key_share" extension with the server's ephemeral
+Diffie-Hellman share which MUST be in the same group.
+If a pure PSK or an (EC)DHE-PSK cipher suite is negotiated, then the
+ServerHello will contain a "pre_shared_key" extension indicating
+which if the client's offered PSKs was selected.
```

Comment:
I like how this works out:

if (ECDHE) add(key_share)
if (PSK) add(pre_shared_key)